### PR TITLE
Add "wxkg" Tuya Smart Button

### PR DESCRIPTION
**Pull request by wilwac**
## Description

Adds smart button - last pushed button sequence (click, double_click or press) will show in HA.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

With love

- [x] Local Testing on HA with Tuya Smart buttons

## Diagnostic File

N/A

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by brucemiranda** [comment]
@wilwac if Button 1 has been Clicked and then a little while later, Button 1 is Clicked again, how will HA report that, since there won't be a change of State between the first Click and the second Click.

**Comment by habitats-tech** [comment]
HA tracks events which can have a defined state or not. Therefore however many times a button is pressed HA will process the event with whatever state. Simply put, HA does not track states it reports states when an event takes place.

https://www.home-assistant.io/docs/configuration/events/

**Comment by brucemiranda** [comment]
> HA tracks events which can have a defined state or not. Therefore however many times a button is pressed HA will process the event with whatever state. Simply put, HA does not track states it reports states when an event takes place.
> 
> https://www.home-assistant.io/docs/configuration/events/

So this seem to suggest that a state_changed event will be triggered even if the previous state and new state havn't actually changed?

**Comment by brucemiranda** [comment]
@shihuimiao how long until this PR gets merged?

**Comment by habitats-tech** [comment]
> > HA tracks events which can have a defined state or not. Therefore however many times a button is pressed HA will process the event with whatever state. Simply put, HA does not track states it reports states when an event takes place.
> > https://www.home-assistant.io/docs/configuration/events/
> 
> So this seem to suggest that a state_changed event will be triggered even if the previous state and new state havn't actually changed?

State change logic must be created/handled/processed by you. HA just reports the states. How you process state changes is up to you (see automations and by extension blueprints, scripts, scenes).

**Comment by brucemiranda** [comment]
> > > HA tracks events which can have a defined state or not. Therefore however many times a button is pressed HA will process the event with whatever state. Simply put, HA does not track states it reports states when an event takes place.
> > > https://www.home-assistant.io/docs/configuration/events/
> > 
> > 
> > So this seem to suggest that a state_changed event will be triggered even if the previous state and new state havn't actually changed?
> 
> State change logic must be created/handled/processed by you. HA just reports the states. How you process state changes is up to you (see automations and by extension blueprints, scripts, scenes).

I get that, I just didn't know that HA would trigger a state_changed event even though the state of the sensor won't have actually changed from its previous value.
I have ESPHome sensors that don't update in HA when HA detects that the new sensor state that ESPHome is trying to update is the same as the state value that is already in HA. Hence my questions. I'll now wait for the PR to get merged into the main branch and then test it out. 
I can then finally get rid of the Virtual switch device that I had created to workaround this unsupported Scene Switch.

**Comment by habitats-tech** [comment]
Just to add a bit of clarity. The event is button pressed. The state of the button pressed event is logged in HA but not processed. What is processed is button pressed, the event itself. How you process a state change is up to you, but in HA nomenclature a state change is not an event (the event is button pressed).

**Comment by habitats-tech** [comment]
For additional clarity please read this: https://developers.home-assistant.io/docs/integration_events

**Comment by brucemiranda** [comment]
I have all my Automations triggered when a sensor State changes. Maybe I could have used the event instead.

**Comment by habitats-tech** [comment]
State change is the right way to automate things. Event trigger is useful sometimes, but generally you want to process when the state changes, not when the event triggers. There is no point activating a light when it is already activated, unless you want to make certain if an event did not actually trigger a state, next time it happens; although you can use a HACS integration for this (Retry), or process the event regardless of state (powering on an already powered on light is of no consequence, except CPU cycles).

**Comment by shihuimiao** [comment]
> @shihuimiao how long until this PR gets merged?

We will complete the processing this week

**Comment by wilwac** [comment]
Thanks all for the discussion. Am entirely new to Home Assistant so finding this conversation helpful. Would be good to hear if this works for you @brucemiranda

File: custom_components/smartlife/sensor.py [hunk 1]
```diff
@@ -1046,6 +1046,56 @@ class SmartLifeSensorEntityDescription(SensorEntityDescription):
             state_class=SensorStateClass.MEASUREMENT,
         ),
     ),
+    # Button or Multi-button device
+    # https://developer.tuya.com/en/docs/iot/f?id=Kbeoa30s4fcdf
+    "wxkg": (
+        SmartLifeSensorEntityDescription(
+            key=DPCode.SWITCH_MODE_1,
+            name="Click",
```
Comment (shihuimiao) [inline custom_components/smartlife/sensor.py]: You can take a look at the app display and suggest using Switch1

File: custom_components/smartlife/sensor.py [hunk 2]
```diff
@@ -1046,6 +1046,56 @@ class SmartLifeSensorEntityDescription(SensorEntityDescription):
...
+            icon="mdi:gesture-tap-button",
+        ),
+        SmartLifeSensorEntityDescription(
+            key=DPCode.SWITCH_MODE_2,
+            name="Click",
```
Comment (shihuimiao) [inline custom_components/smartlife/sensor.py]: using Switch2, The same goes for the following.

File: custom_components/smartlife/sensor.py (hunk 1 above)
Comment (shihuimiao) [reply custom_components/smartlife/sensor.py]: I mean the name attribute.

Comment (brucemiranda) [reply custom_components/smartlife/sensor.py]: @wilwac I think this is a valid comment and it's the way the wkcz and kg categories have been coded in the switch.py file. So having unique names for the Switch Mode sensors does make sense.

**Comment by wilwac** [comment]
agree with comments, changed the name attribute to something more appropriate as per the format of equivalents in switch.py

**Comment by niekp** [comment]
I've tested this adds the buttons as sensor with a 'click' state. When i press the button nothing changes so automations on this 'Button 1' entity don't do anything.

How did you test this? @wilwac @shihuimiao

**Comment by brucemiranda** [comment]
> I've tested this adds the buttons as sensor with a 'click' state. When i press the button nothing changes so automations on this 'Button 1' entity don't do anything.
> 
> How did you test this? @wilwac @shihuimiao

Yes this is what I feared would happen. If the state of the sensor doesn't change because the previous state and the new state are the same, then you cannot Automate anything in HA based on a state change, because it doesn't.
The way Shelly gets around this is by offering its own Automation trigger which gets fired whenever any event takes place. So you are not having to track the state change of the sensor but simply act on the Automation trigger fired. 

The only way around this would be to set the State of the Sensor to Unknown via an Automation after you process the click event.

**Comment by brucemiranda** [comment]
So far this is how I have solved this issue.

At the end of the Automation I forced the sensor state back to unknown. So that a new click event is possible at the next click. 

However I have no idea after how long will the smartlife integration refresh the states of all sensors. 

And even this solution has a problem if you have multiple buttons because any button clicked causes all button states to be refreshed. So if Button 1 has been reset to unknown within HA but when Button 2 is clicked the integration resets Button 1 status back to click because that is what is the last status of that button held in the Tuya cloud.

```
trigger:
  - platform: state
    entity_id:
      - sensor.smartlife_scene_switch_1_button_1
    to: click
condition: []
action:
  - service: python_script.set_state
    data_template:
      entity_id: sensor.smartlife_scene_switch_1_button_1
      state: unknown
```

**Comment by brucemiranda** [comment]
> HA tracks events which can have a defined state or not. Therefore however many times a button is pressed HA will process the event with whatever state. Simply put, HA does not track states it reports states when an event takes place.
> 
> https://www.home-assistant.io/docs/configuration/events/

@habitats-tech this doesn't seem to be working for me. Either HA or smartlife don't seem to be processing or triggering an event when the state of the sensor is the same. So a new button click doesn't raise any HA event if the old state is click and the new state is click as well.

**Comment by habitats-tech** [comment]
Click events are like black magic. It could be that the product vendor at firmware level or Tuya at API/app level do not allow click events to be registered within a time period and therefore never reach HA. It could be HA is doing this, although I doubt is the case. All events are handled by the event bus, which is an area you might want to research. Do the events reach HA? If you are confident they do share your reasoning behind it to see if I can assist.

What is the time interval of two consecutive clicks is critical to what you do. Almost everyone in the industry will not allow click events to take place in a very short timespan.

**Comment by brucemiranda** [comment]
> Click events are like black magic. It could be that the product vendor at firmware level or Tuya at API/app level do not allow click events to be registered within a time period and therefore never reach HA. It could be HA is doing this, although I doubt is the case. All events are handled by the event bus, which is an area you might want to research. Do the events reach HA? If you are confident they do share your reasoning behind it to see if I can assist.
> 
> What is the time interval of two consecutive clicks is critical to what you do. Almost everyone in the industry will not allow click events to take place in a very short timespan.

The click event is reaching HA. The issue is that if the state of the sensor is already click, then then next click is being completely ignored by HA. It doesn't even raise a state_changed event because according to HA the state hasn't changed. 
If you have a single button then you can work around this by forcing the HA sensor state to unknown after you process the click event. However if you have a multiple button switch then all button states are updated with any button click.

**Comment by habitats-tech** [comment]
Interesting. Will try to test next month and see if we can come up with workarounds or fixes.

**Comment by habitats-tech** [comment]
The behaviour you are suggesting is a bit strange. I am hesitant to believe HA will reject a click event as the same button might have toggle functionality which HA processes. Things do not make to much sense.

**Comment by brucemiranda** [comment]
> Interesting. Will try to test next month and see if we can come up with workarounds or fixes.

I saw this same behaviour when ESPHome pushes out updates to sensors within HA. HA does not update the sensor when the current value of the sensor is the same as the update. So it looks as though ESPHome hasn't updated the sensor within HA.

**Comment by habitats-tech** [comment]
Hmmm.... I need to test this. Unfortunately earliest opportunity will only be next month.

**Comment by brucemiranda** [comment]
> The behaviour you are suggesting is a bit strange. I am hesitant to believe HA will reject a click event as the same button might have toggle functionality which HA processes. Things do not make to much sense.

A new state is processed absolutely fine. So a click state changes to press on a long button press and back to click on a short button click.  However if the state is press and you long press the button again, HA does not process that state as a change. Same with click too.

**Comment by habitats-tech** [comment]
Just to understand what you are suggesting. An automation listening to the button pressed event will not trigger every time the button is pressed?

**Comment by brucemiranda** [comment]
> Just to understand what you are suggesting. An automation listening to the button pressed event will not trigger every time the button is pressed?

Correct. Because my Automation is looking for a state_changed event which never fires. I even used the Developer Tools to see if there were any other events fired when I pressed a button and there are None. An event is fired only if the state changes I.e. If the button is pressed differently from the last time it was pressed.

**Comment by brucemiranda** [comment]
I think maybe the issue here is that the integration can only handle the return status that is coming back from the Tuya cloud. It doesn't seem to be able to trap the Report event that the button raises to the Tuya cloud when a button is pressed.

**Comment by habitats-tech** [comment]
Interesting. I need to clear some of my backlog and will delve into this later as what you are implying requires some TLC.

**Comment by nicktrigger** [comment]
Hi, I see this is merged, but I still have my smart zigbee m4 switch showing as unsupported with no entities. The diagnostics file shows all the buttons and modes for each button. Diagnostics attached...

[tuya-d28d44a56af41a62febd2ef2e6830422-Living Room Scene Switch-fa5b1f50bb75543ac7850a7805221d82.json](https://github.com/tuya/tuya-smart-life/files/14231110/tuya-d28d44a56af41a62febd2ef2e6830422-Living.Room.Scene.Switch-fa5b1f50bb75543ac7850a7805221d82.json)

**Comment by brucemiranda** [comment]
> Hi, I see this is merged, but I still have my smart zigbee m4 switch showing as unsupported with no entities. The diagnostics file shows all the buttons and modes for each button. Diagnostics attached...
> 
> [tuya-d28d44a56af41a62febd2ef2e6830422-Living Room Scene Switch-fa5b1f50bb75543ac7850a7805221d82.json](https://github.com/tuya/tuya-smart-life/files/14231110/tuya-d28d44a56af41a62febd2ef2e6830422-Living.Room.Scene.Switch-fa5b1f50bb75543ac7850a7805221d82.json)

The reason you see this as unsupported is because although your switch has the same category **wxkg** your switches themselves are coming in as  **switch1_value**  rather than **switch_mode1** 
So the code will need to be changed to add support for your particular switch values in addition to switch modes.

**Comment by nicktrigger** [comment]
I see, thank you for the clarification. Unfortunately I'm not really familiar with coding. Hopefully someone will pick it up and add this to the code!

**Comment by brucemiranda** [comment]
> I see, thank you for the clarification. Unfortunately I'm not really familiar with coding. Hopefully someone will pick it up and add this to the code!

The way it is currently coded, it is of little use anyway. So you aren't missing much. HA cannot detect the same switch mode twice. So is useless for Automations.

**Pull request merged**